services/QuillLMS/app/services/diagnostic_insights/activity_session_skill_performance.rb

```diff
@@ -0,0 +1,116 @@
+# frozen_string_literal: true
+
+module DiagnosticInsights
+  class ActivitySessionSkillPerformance < ApplicationService
+    attr_reader :activity_session
+
+    def initialize(activity_session)
+      @activity_session = activity_session
+    end
+
+    def run
+      return unless concept_results.present?
+      return unless diagnostic? || recommendation?
+
+      skill_performance
+    end
+
+    private def skill_performance
+      # NOTE: Because we've never consolidated these, "Skills" for Diagnostic activities are derived from the "SkillGroup" model,
+      # while "Skills" for recommended activities are derived from the "Recommendation" model.
+      # There is no direct database mapping linking these together, just the intentional coincidence of names
+      questions_correct
+        .group_by { |question_uid, _| skill_group_for_question(question_uid) }
+        .map do |skill, performance|
+          {
+            skill_id: skill.id,
+            skill_name: skill.name,
+            questions_total: performance.length,
+            questions_correct: performance.sum { |question_performance| question_performance.second ? 1 : 0 }
+          }
+        end
+    end
+
+    private def questions_correct
+      return diagnostic_questions_correct if diagnostic?
+
+      recommendation_questions_correct if recommendation?
+    end
+
+    private def diagnostic_questions_correct
+      concept_results
+        .group_by(&:question_uid)
+        .transform_values do |attempts|
+          attempts.select do |attempt|
+            optimal_concepts = questions[attempt.question_uid].diagnostic_question_optimal_concepts
+            attempt.correct && (
+              optimal_concepts.empty? ||
+              optimal_concepts.pluck(:concept_id).include?(attempt.concept_id))
+          end.any?
+        end
+    end
+
+    private def recommendation_questions_correct
+      concept_results
+        .group_by(&:question_uid)
+        .transform_values do |attempts|
+          last_attempt = attempts.map(&:attempt_number).max
+          attempts.select { |attempt| attempt.attempt_number == last_attempt }
+            .map(&:correct)
+            .all?
+        end
+    end
+
+    private def activity = activity_session&.activity
+    private def activity_id = activity&.id
+    private def activity_session_id = activity_session&.id
+    private def classroom_unit = activity_session&.classroom_unit
+    private def concept_results = activity_session&.concept_results
+    private def diagnostic? = Activity::ALL_PRE_TEST_DIAGNOSTIC_IDS.include?(activity_id) || Activity::ALL_POST_TEST_DIAGNOSTIC_IDS.include?(activity_id)
+    private def question_uids = concept_results.map(&:question_uid)
+    private def recommendation? = classroom_unit && recommendation
+
+    private def skill_groups
```

Comment:
I think it would be worth tailing the develop log while running this job since you're not going through active_record associations (you're doing each model query individually) to see the SQL generated. There may be a more optimized way to do this (it's hard to see just with the code).